### PR TITLE
`simulationResultsToDataFrame()` speed up

### PR DESCRIPTION
**Pull request by PavelBal**

### CODE REVIEW

File: R/utilities-simulation-results.R [hunk 1]
```diff
@@ -223,59 +223,30 @@ simulationResultsToDataFrame <- function(simulationResults,
...
-    dplyr::tibble(
-      "TimeUnit" = simList$metaData$Time$unit[[1]],
-      "TimeDimension" = simList$metaData$Time$dimension[[1]]
+  units <- lapply(simList$metaData, function(path) {
+    path$unit
+  })
+  dims <- lapply(simList$metaData, function(path) {
+    path$dimension
+  })
+  molWeights <- lapply(unique(simData$paths), function(path) {
```
Comment (PavelBal) [inline R/utilities-simulation-results.R:232]: @Felixmil any better way to do this?

File: R/utilities-simulation-results.R [hunk 2]
```diff
@@ -223,59 +223,30 @@ simulationResultsToDataFrame <- function(simulationResults,
...
-  # with `nest()`, the computation only be carried for the same number of
-  # times as the number of `paths` present.
-  simData <- simData %>%
-    dplyr::group_by(paths) %>%
-    tidyr::nest() %>%
-    # Add a new column for molecular weight.
-    # When you call `molWeightFor()`, it returns the value in the base unit -
-    # which is `kg/µmol`. This is not the unit the user would expect, so we
-    # convert it first to the common unit `g/mol`.
+  simData <- dplyr::group_by(simData, paths) %>%
```
Comment (PavelBal) [inline R/utilities-simulation-results.R:241]: @Felixmil Also here - any speed up ideas?

File: tests/testthat/test-utilities-simulation-results.R [hunk 3]
```diff
@@ -166,12 +166,12 @@ test_that("simulationResultsToDataFrame works as expected - minimal pkml", {
   expect_equal(unique(df2$paths), "Organism|A")
 
   # names
-  expect_equal(
-    names(df1),
-    c(
-      "paths", "IndividualId", "Time", "simulationValues", "unit",
-      "dimension", "TimeUnit", "TimeDimension", "molWeight"
-    )
+  expect_true(
```
Comment (PavelBal) [inline tests/testthat/test-utilities-simulation-results.R:169]: Changed the test so it does not test for the exact order of columns. The order has changed to 

`"IndividualId"     "Time"             "paths"            "simulationValues" "TimeDimension"    "TimeUnit"         "dimension"        "unit"             "molWeight"  `

**Comment by PavelBal** [comment]
Fixes #1287

File: R/utilities-simulation-results.R (hunk 1 above)
Comment (Felixmil) [reply R/utilities-simulation-results.R:232]: This should be better performance and readability wise:

```r
units <- purrr::map(simList$metaData, "unit")

dims <- purrr::map(simList$metaData, "dimension")

molWeights <- unique(simData$paths) %>%
  purrr::set_names() %>%
  purrr::map(~ospsuite::toUnit(
    quantityOrDimension = ospDimensions$`Molecular weight`,
    values              = simulationResults$simulation$molWeightFor(.x),
    targetUnit          = ospUnits$`Molecular weight`$`g/mol`
  ))
```
\+ remove `names(molWeights) <- unique(simData$paths)`

File: R/utilities-simulation-results.R (hunk 2 above)
Comment (Felixmil) [reply R/utilities-simulation-results.R:241]: this tibble operation should not be very long, however we can still improve it a bit for this kind of column mutation using `rowwise`:
```r
  simData <-
    simData %>%
    dplyr::rowwise() %>%
    dplyr::mutate(
      TimeDimension = dims$Time,
      TimeUnit = units$Time,
      dimension = dims[[paths]],
      unit = units[[paths]],
      molWeight = molWeights[[paths]]
    )
```

File: R/utilities-simulation-results.R (hunk 1 above)
Comment (PavelBal) [reply R/utilities-simulation-results.R:232]: Yepp a bit better.

File: R/utilities-simulation-results.R (hunk 2 above)
Comment (PavelBal) [reply R/utilities-simulation-results.R:241]: This actually DOUBLED the time needed :D

File: tests/testthat/test-utilities-simulation-results.R (hunk 3 above)
Comment (svavil) [reply tests/testthat/test-utilities-simulation-results.R:169]: This tests that there are no additional columns in `names(df1)`, but not that all columns are present. Maybe 
```
expect_true(all(names(df1) %in% c(...))
expect_true(all(c(...) %in% names(df1))
```
or
```
expect_equal(sort(names(df)), sort(c(...)))
```

File: R/utilities-simulation-results.R (hunk 2 above)
Comment (Felixmil) [reply R/utilities-simulation-results.R:237]: Here is the solution using `{data.table}`:

```r
  simData <-  data.table::as.data.table(simData)
  simData <- simData[,`:=`(TimeDimension =  dims$Time,
                           TimeUnit =  units$Time,
                           dimension =  dims[[paths]],
                           unit =  units[[paths]],
                           molWeight = molWeights[[paths]]),
                     by = paths]
```
Result: ~2 times faster. This should be even faster with large datasets
```
Unit: milliseconds
                                     expr    min      lq     mean median      uq     max neval
     og(simData, dims, units, molWeights) 4.3466 4.89990 5.936743 5.0822 5.40190 22.0154   100
    new(simData, dims, units, molWeights) 3.9706 4.34070 5.003569 4.5498 4.91425 18.0941   100
 new_dt(simData, dims, units, molWeights) 1.4282 2.07735 3.029173 2.3516 2.65680 16.4108   100
```

Comment (Felixmil) [reply R/utilities-simulation-results.R:237]: Results with a simData with `2 946 000` rows:
```r
Unit: milliseconds
                                         expr      min       lq     mean   median       uq      max neval
     og(big_simData, dims, units, molWeights) 204.5205 219.9517 326.0187 322.4694 430.7714 449.7938    10
    new(big_simData, dims, units, molWeights) 157.8323 180.3917 232.1167 187.7086 320.0275 374.1119    10
 new_dt(big_simData, dims, units, molWeights) 101.9321 119.9672 180.7313 123.3568 303.0376 340.7760    10
```

**Review by Felixmil — COMMENTED** [review]
implement either new dplyr or data.table solutions

**Comment by PavelBal** [comment]
@Felixmil nice, feel free to push.

**Comment by PavelBal** [comment]
Hmm but the reference performance will be different for each machine?

**Comment by svavil** [comment]
> Hmm but the reference performance will be different for each machine?

@PavelBal that's why I'm setting a 2-fold margin (maybe set a 3-fold or something), but basically, I don't want the performance to suddenly become worse in a year because some changes were accumulating.

**Review by Felixmil — APPROVED** [review]

**Comment by Yuri05** [comment]
> > Hmm but the reference performance will be different for each machine?
> 
> @PavelBal that's why I'm setting a 2-fold margin (maybe set a 3-fold or something), but basically, I don't want the performance to suddenly become worse in a year because some changes were accumulating.

 If we add benchmark tests, I suggest to define a precondition which is FALSE as per default.
E.g. "benchmark test starts only if environment variable OSPSUITE-BENCHMARK-TESTS is defined".

**Comment by svavil** [comment]
@PavelBal we can make it more machine-agnostic, if you are fine with some arbitrary benchmarks, for example: 
```
sim <- loadTestSimulation("MinimalModel")
simResults <- importResultsFromCSV(
  simulation = sim,
  filePaths = getTestDataFilePath("Stevens_2012_placebo_indiv_results.csv")
)
# This is an example benchmark that takes around 4 seconds
benchmark_1e8 <- system.time({for (i in 1:1e8) {res <- 2**0.5}})
# In the current implementation, 100 runs of simulationResultsToDataFrame work
# approximately twice faster than the benchmark above
test_that("The simulationResultsToDataFrame() performance is not degraded, compared to a reference performance", {
  benchmark <- system.time({
    for (i in 1:100) {
      simulationResultsToDataFrame(simResults)
    }
  })
  expect_lt(benchmark[["elapsed"]], benchmark_1e8[["elapsed"]])
})
```

Of course, wrapped in a flag that Yuri proposed.

**Comment by PavelBal** [comment]
@svavil Feel free to make a PR for benchmark tests. In the meantime, I am merging this.

**Pull request merged**